File: machine/cpp/core/host_frame.cpp
#include "core/machine_manager.h"

#include "core/host_overlay_menu.h"
#include "common/time.h"
#include "input/manager.h"
#include "machine/runtime/runtime.h"
#include "render/video_presenter.h"

#include <chrono>

namespace bmsx {
namespace {
constexpr double MAX_FRAME_DELTA_MS = 250.0;
}

bool MachineManager::runHostFrame(
	Runtime& runtime,
	MicrotaskQueue& microtasks,
	VideoPresenter& presenter,
	f64 deltaTime,
	bool platformPaused
) {
	if (!acceptHostFrame(deltaTime)) {
		return false;
	}
	const auto tickStart = std::chrono::steady_clock::now();
	m_screen.recordHostFrame();

	double hostDeltaMs = deltaTime * 1000.0;
	if (hostDeltaMs > MAX_FRAME_DELTA_MS) {
		hostDeltaMs = MAX_FRAME_DELTA_MS;
	}
	const double hostDeltaSeconds = hostDeltaMs / 1000.0;
	m_delta_time = hostDeltaSeconds;
	m_total_time += hostDeltaSeconds;
	m_frame_count += 1;
	runtime.frameLoop.currentTimeSeconds = m_total_time;
	m_fps = 1.0 / hostDeltaSeconds;

	Input::instance().pollInput();
	const bool hostMenuActive = hostOverlayMenu().tickInput(*this, presenter);

	m_screen.clearPresentation();
	if (!platformPaused && !hostMenuActive) {
		m_delta_time = runtime.timing.frameDurationMs / 1000.0;
		const i64 previousTickSequence = runtime.frameScheduler.lastTickSequence;
		runtime.frameScheduler.run(runtime, hostDeltaMs);
		while (runtime.machine.gxGpu.backendReadbackPending()) {
			presenter.backend().executeGxGpuReadback(runtime.machine.gxGpu);
			runtime.frameScheduler.run(runtime, 0.0);
		}
		syncRuntimeAudioTiming();
		m_screen.syncAfterRuntimeUpdate(runtime, previousTickSequence);
	} else {
		runtime.frameScheduler.clearQueuedTime();
	}
	m_delta_time = hostDeltaSeconds;

	microtasks.flush();

	m_last_tick_timing.totalMs = to_ms(std::chrono::steady_clock::now() - tickStart);

	if (hostMenuActive) {
		hostOverlayMenu().queueRenderCommands(*this, presenter);
		m_screen.requestHeldPresentation();
	} else if (hostOverlayMenu().queueFrameOverlayCommands(*this, presenter)) {
		m_screen.requestHeldPresentation();
	}
	return m_screen.render(*this, presenter, runtime, platformPaused);
}

} // namespace bmsx

File: machine/ts/core/host_frame.ts
import { machineManager } from './machine_manager';
import { hostOverlayMenu } from './host_overlay_menu';
import * as workbenchMode from '../ide/workbench/mode';
import { flushRuntimeLuaOutputToTerminal } from '../ide/runtime/state';
import type { Runtime } from '../machine/runtime/runtime';
import { runRuntimeFrameStepInto, type RuntimeFrameStepResult } from '../machine/runtime/frame/step';

const MAX_HOST_FRAME_DELTA_MS = 250;
const hostFrameStepResult: RuntimeFrameStepResult = {
	previousTickSequence: 0,
	tickSequence: 0,
	tickAdvanced: false,
};

export function runMachineHostFrame(runtime: Runtime, currentTime: number, runReady: boolean): void {
	const manager = machineManager;
	if (!manager.running) {
		return;
	}
	const screen = manager.screen;
	let hostDeltaMs = 0;
	try {
		manager.input.pollInput();
		screen.beginHostFrame(currentTime);
		workbenchMode.tickIdeInput();
		workbenchMode.tickTerminalInput(runtime);
		hostDeltaMs = Math.min(currentTime - runtime.frameLoop.currentTimeMs, MAX_HOST_FRAME_DELTA_MS);
		runtime.frameLoop.currentTimeMs = currentTime;
		manager.host_fps = 1000 / hostDeltaMs;
		const hostMenuActive = hostOverlayMenu.tickInput();

		if (hostMenuActive) {
			screen.clearPresentation();
			runtime.frameScheduler.clearQueuedTime();
			hostOverlayMenu.queueRenderCommands();
			screen.requestHeldPresentation();
			manager.platform.microtasks.flush();
			screen.presentPending(runtime, hostDeltaMs);
		} else if (manager.paused) {
			hostOverlayMenu.queueFrameOverlayCommands();
			manager.platform.microtasks.flush();
			screen.presentPausedFrame(runtime, hostDeltaMs);
		} else {
			const hostOverlayQueued = hostOverlayMenu.queueFrameOverlayCommands();
			screen.clearPresentation();
			if (runtime.executionOverlayActive) {
				screen.runOverlay(runtime);
			} else if (!runReady) {
				runtime.frameScheduler.clearQueuedTime();
			} else {
				manager.deltatime = runtime.timing.frameDurationMs;
				runRuntimeFrameStepInto(hostFrameStepResult, runtime, hostDeltaMs);
				manager.syncRuntimeAudioTiming();
				screen.syncAfterRuntimeUpdate(runtime, hostFrameStepResult.previousTickSequence);
			}
			if (hostOverlayQueued) {
				screen.requestHeldPresentation();
			}
			manager.platform.microtasks.flush();
			screen.presentPending(runtime, hostDeltaMs);
		}
	} catch (error) {
		workbenchMode.surfaceHostFrameError(runtime, error, hostDeltaMs, screen);
	}
	flushRuntimeLuaOutputToTerminal(runtime, manager.ideState);
	screen.flushDebugReport(currentTime, runtime);
}
